src/main/java/net/snowflake/ingest/streaming/internal/BlobBuilder.java

```diff
@@ -48,6 +56,106 @@ class BlobBuilder {
 
+  /**
+   * Builds blob.
+   *
+   * @param filePath Path of the destination file in cloud storage
+   * @param blobData All the data for one blob. Assumes that all ChannelData in the inner List
+   *     belongs to the same table. Will error if this is not the case
+   * @param bdecVersion version of blob
+   * @return {@link Blob} data
+   */
+  static <T> Blob build(
+      String filePath, List<List<ChannelData<T>>> blobData, Constants.BdecVersion bdecVersion)
+      throws IOException, NoSuchPaddingException, NoSuchAlgorithmException,
+          InvalidAlgorithmParameterException, InvalidKeyException, IllegalBlockSizeException,
+          BadPaddingException {
+    List<ChunkMetadata> chunksMetadataList = new ArrayList<>();
+    List<byte[]> chunksDataList = new ArrayList<>();
+    long curDataSize = 0L;
+    CRC32 crc = new CRC32();
+
+    // TODO: channels with different schema can't be combined even if they belongs to same table
+    for (List<ChannelData<T>> channelsDataPerTable : blobData) {
+      ByteArrayOutputStream chunkData = new ByteArrayOutputStream();
+      SnowflakeStreamingIngestChannelInternal.ChannelContext firstChannelContext =
+          channelsDataPerTable.get(0).getChannelContext();
+
+      Flusher<T> flusher = channelsDataPerTable.get(0).createFlusher();
+      Flusher.SerializationResult result =
+          flusher.serialize(channelsDataPerTable, chunkData, filePath);
+
+      if (!result.channelsMetadataList.isEmpty()) {
+        Pair<byte[], Integer> compressionResult =
+            compressIfNeededAndPadChunk(
+                filePath,
+                chunkData,
+                Constants.ENCRYPTION_ALGORITHM_BLOCK_SIZE_BYTES,
+                bdecVersion == Constants.BdecVersion.ONE);
+        byte[] compressedAndPaddedChunkData = compressionResult.getFirst();
+        int compressedChunkLength = compressionResult.getSecond();
+
+        // Encrypt the compressed chunk data, the encryption key is derived using the key from
+        // server with the full blob path.
+        // We need to maintain IV as a block counter for the whole file, even interleaved,
+        // to align with decryption on the Snowflake query path.
+        // TODO: address alignment for the header SNOW-557866
+        long iv = curDataSize / Constants.ENCRYPTION_ALGORITHM_BLOCK_SIZE_BYTES;
+        byte[] encryptedCompressedChunkData =
+            Cryptor.encrypt(
+                compressedAndPaddedChunkData, firstChannelContext.encryptionKey, filePath, iv);
+
+        // Compute the md5 of the chunk data
+        String md5 = computeMD5(encryptedCompressedChunkData, compressedChunkLength);
+        int encryptedCompressedChunkDataSize = encryptedCompressedChunkData.length;
+
+        // Create chunk metadata
+        long startOffset = curDataSize;
+        ChunkMetadata chunkMetadata =
+            ChunkMetadata.builder()
+                .setOwningTable(firstChannelContext)
```

Comment:
This line sounds weird. Either `.setOwningChannel(firstChannelContext)` or `.setOwingTable(firstChannelContext.getTable)` would make sense.